clang/lib/DPCT/AnalysisInfo.cpp

```diff
@@ -2631,7 +2631,8 @@ void DeviceFunctionDecl::LinkDecl(const FunctionDecl *FD, DeclList &List,
       TSK_ExplicitInstantiationDefinition)
     return;
   if (FD->isImplicit() ||
-      FD->getTemplateSpecializationKind() == TSK_ImplicitInstantiation) {
+      (FD->getTemplateSpecializationKind() == TSK_ImplicitInstantiation &&
```

Comment:
Please refer to the code snippet in https://godbolt.org/z/P95YWEE4o

`FD` could be a non-template member function of local class(e.g., `MaxOp::warp_shfl_down`) that is defined inside CUDA kernel template function. 
`FD->getTemplateSpecializationKind() == TSK_ImplicitInstantiation` is `true` in such case, but `FD` itself is not template function instantiation.

Reply:
FunctionDecl::getPrimaryTemplate()
https://clang.llvm.org/doxygen/classclang_1_1FunctionDecl.html#a3701fbf27221163b5f533f7d0a6f418c